File: implants/lib/portals/portal-stream/src/reader.rs
use anyhow::{Result, anyhow};
use pb::portal::Mote;
use std::collections::BTreeMap;
use std::time::{Duration, Instant};

/// OrderedReader receives Motes and ensures they are read in order.
pub struct OrderedReader {
    next_seq_id: u64,
    buffer: BTreeMap<u64, Mote>,
    max_buffer: usize,
    stale_timeout: Duration,
    first_buffered_at: Option<Instant>,
}

impl OrderedReader {
    /// Creates a new OrderedReader with default settings.
    pub fn new() -> Self {
        Self {
            next_seq_id: 0,
            buffer: BTreeMap::new(),
            max_buffer: 1024,
            stale_timeout: Duration::from_secs(5),
            first_buffered_at: None,
        }
    }

    /// Sets the maximum number of out-of-order messages to buffer.
    pub fn with_max_buffered_messages(mut self, max: usize) -> Self {
        self.max_buffer = max;
        self
    }

    /// Sets the duration to wait for the next expected sequence ID before erroring.
    pub fn with_stale_buffer_timeout(mut self, timeout: Duration) -> Self {
        self.stale_timeout = timeout;
        self
    }

    /// Processes a new mote.
    /// Returns:
    /// - Ok(Some(vec![...])): A list of ordered motes ready to be consumed.
    /// - Ok(None): The mote was buffered (gap detected) or dropped (duplicate).
    /// - Err(e): Buffer limit exceeded or stale timeout detected *during* processing.
    pub fn process(&mut self, mote: Mote) -> Result<Option<Vec<Mote>>> {
        // Check for duplicate/old packet
        if mote.seq_id < self.next_seq_id {
            return Ok(None);
        }

        // Check if this is the expected packet
        if mote.seq_id == self.next_seq_id {
            let mut result = Vec::new();
            result.push(mote);
            self.next_seq_id += 1;

            // Check if we have subsequent packets in the buffer
            while let Some(next_mote) = self.buffer.remove(&self.next_seq_id) {
                result.push(next_mote);
                self.next_seq_id += 1;
            }

            // Update timer based on buffer state
            if self.buffer.is_empty() {
                self.first_buffered_at = None;
            } else {
                // Buffer still has items (gaps further ahead), so we are technically waiting for the next one.
                // Resetting timer to give full timeout for the next expected packet.
                self.first_buffered_at = Some(Instant::now());
            }

            return Ok(Some(result));
        }

        // Gap detected: mote.seq_id > self.next_seq_id
        if self.buffer.is_empty() {
            self.first_buffered_at = Some(Instant::now());
        }

        self.buffer.entry(mote.seq_id).or_insert(mote);

        if self.buffer.len() > self.max_buffer {
            return Err(anyhow!("stale stream: buffer limit exceeded"));
        }

        // We check timeout here as well, similar to Go implementation
        self.check_timeout()?;

        Ok(None)
    }

    /// Checks if the reader has stalled waiting for a packet.
    pub fn check_timeout(&self) -> Result<()> {
        if !self.buffer.is_empty()
            && let Some(start) = self.first_buffered_at
            && start.elapsed() > self.stale_timeout
        {
            return Err(anyhow!(
                "stale stream: timeout waiting for seqID {}",
                self.next_seq_id
            ));
        }
        Ok(())
    }
}

impl Default for OrderedReader {
    fn default() -> Self {
        Self::new()
    }
}

File: tavern/portals/stream/reader.go
package stream

import (
	"errors"
	"fmt"
	"time"

	"realm.pub/tavern/portals/portalpb"
)

// ReceiverFunc is a callback that reads a Mote from a source.
// This allows OrderedReader to wrap any gRPC stream method.
type ReceiverFunc func() (*portalpb.Mote, error)

// OrderedReader receives Motes and ensures they are read in order.
type OrderedReader struct {
	nextSeqID       uint64
	buffer          map[uint64]*portalpb.Mote
	maxBuffer       int
	staleTimeout    time.Duration
	firstBufferedAt time.Time
}

// WithMaxBufferedMessages sets the maximum number of out-of-order messages to buffer.
func WithMaxBufferedMessages(max int) func(*OrderedReader) {
	return func(r *OrderedReader) {
		r.maxBuffer = max
	}
}

// WithStaleBufferTimeout sets the duration to wait for the next expected sequence ID before erroring if other messages are arriving.
func WithStaleBufferTimeout(d time.Duration) func(*OrderedReader) {
	return func(r *OrderedReader) {
		r.staleTimeout = d
	}
}

// NewOrderedReader creates a new OrderedReader.
// maxBuffer limits the number of out-of-order messages to buffer.
// staleTimeout is the duration to wait for the next expected sequence ID before erroring if other messages are arriving.
func NewOrderedReader(options ...func(*OrderedReader)) *OrderedReader {
	reader := &OrderedReader{
		nextSeqID:    0,
		buffer:       make(map[uint64]*portalpb.Mote),
		maxBuffer:    1024,
		staleTimeout: 5 * time.Second,
	}
	for _, opt := range options {
		opt(reader)
	}
	return reader
}

// Process handles a new incoming mote.
// If the mote is the expected next one, it returns it along with any subsequent buffered motes that are now in order.
// If the mote is out of order (future), it buffers it and returns nil.
// If the mote is a duplicate (past), it returns nil.
// Returns an error if buffer limits or timeouts are exceeded.
func (r *OrderedReader) Process(mote *portalpb.Mote) ([]*portalpb.Mote, error) {
	if mote.SeqId == r.nextSeqID {
		// Expected packet
		result := []*portalpb.Mote{mote}
		r.nextSeqID++

		// Check for buffered subsequent packets
		for {
			if bufferedMote, ok := r.buffer[r.nextSeqID]; ok {
				result = append(result, bufferedMote)
				delete(r.buffer, r.nextSeqID)
				r.nextSeqID++
			} else {
				break
			}
		}

		// Reset buffer time if buffer becomes empty
		if len(r.buffer) == 0 {
			r.firstBufferedAt = time.Time{}
		} else {
			// Reset timer to now as we made progress
			r.firstBufferedAt = time.Now()
		}

		return result, nil
	} else if mote.SeqId > r.nextSeqID {
		// Future packet (gap)
		if len(r.buffer) == 0 {
			r.firstBufferedAt = time.Now()
		}

		if _, exists := r.buffer[mote.SeqId]; !exists {
			r.buffer[mote.SeqId] = mote
		}

		if len(r.buffer) > r.maxBuffer {
			return nil, errors.New("stale stream: buffer limit exceeded")
		}

		// Check timeout
		if time.Since(r.firstBufferedAt) > r.staleTimeout {
			return nil, fmt.Errorf("stale stream: timeout waiting for seqID %d", r.nextSeqID)
		}

		return nil, nil
	} else {
		// Duplicate or old packet. Ignore.
		return nil, nil
	}
}
